### PR TITLE
Random op python api

### PR DESCRIPTION
**Pull request by jjsjann123**
exposes `normal` & `uniform` on python API. Fixes issue #2386 

A few things that I want to keep working on ( not necessarily in this PR ):
- [ ] Add array support in fusion record state, so each op don't have to wrap up its own support on hash/print/operator(). Currently there are other ops (e.g. `full`) that avoids the problem by not supporting nvfuser::scalar, but uses constant attribute to handle that.
- [ ] We need to discuss how we want to handle python number & nvfuser scalar. Maybe relying on pybind API is a bad idea here, as I don't see how easily we can support both without the tedious boilerplate pybind definition (look at broadcast_in_dim here). I'm thinking about moving these API to python side and have the pybind only takes nvfuser scalar. Our python script will convert python number to nvfuser constant scalar.
^^^ This is just some simple ideas. need to put more thought into this.

### CODE REVIEW

**Comment by jjsjann123** [comment]
I'm leaning towards keeping the uniform and normal PR with a somewhat basic API. 

Currently I'm only taking nvfuser scalar as argument input. You can look at the python test we have. vvv Python numbers are not acceptable for `fd.ops.uniform`
```
        def fusion_func(fd: FusionDefinition) :
            t0 = fd.from_pytorch(inputs[0])
            s_lo = fd.define_constant(lo)
            s_hi = fd.define_constant(hi)
            size = fd.ops.tensor_sizes(t0)
            t1 = fd.ops.uniform(s_lo, s_hi, size, DataType.Double)
            fd.add_output(t1)
```
I want to refactor how we handle python number & nvfuser scalar. The overload done in pybind is too messy... e.g. for uniform. we can have combination of python number and nvfuser scalar, that's going to give us a 2**3 overload !

I'm thinking about pushing the API to python to handle that. But I'll do that in a separate PR cc'ing @kevinstephano

**Comment by jjsjann123** [comment]
I'll deal with python API refactor on separate PR. So I think this one is good for review

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -1523,6 +1523,76 @@ void initNvFuserPythonBindings(PyObject* module) {
       py::arg("correction"),
       py::arg("keepdim") = false,
       py::return_value_policy::reference);
+  nvf_ops.def(
+      "uniform",
+      [](nvfuser::FusionDefinition::Operators& self,
+         nvfuser::Scalar minval,
+         nvfuser::Scalar maxval,
+         std::vector<nvfuser::Scalar>& shape,
```
Comment (jjsjann123) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp]: Note that this one is a little funny and doesn't work look good with the API refactor we had from #2451 

So with dynamic vs static shape, we are currently looking at `Scalar shape` vs `std::vector<nvfuser::Scalar>`, which suggests that we are in the hell with pybind11 overload again... I need to revisit the refactor in #2450 and #2451

**Comment by jjsjann123** [comment]
merged devel back here. This one is good for review again @kevinstephano

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 2]
```diff
@@ -2586,6 +2586,72 @@ void initNvFuserPythonBindings(PyObject* module) {
       py::arg("correction"),
       py::arg("keepdim") = false,
       py::return_value_policy::reference);
+  nvf_ops.def(
```
Comment (jjsjann123) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:2648]: Just want to point out here that, uniform & normal are not tensor methods, so we don't have to repeat it for that.
But the duplication is less than ideal cc'ing @jacobhinkle  :wink:

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 3]
```diff
@@ -2586,6 +2586,72 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+      "uniform",
+      [](FusionDefinition::Operators& self,
+         Scalar minval,
```
Comment (kevinstephano) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:2592]: Was this supposed to have python numbers as min and max values as well for the random factory functions?

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 4]
```diff
@@ -2586,6 +2586,72 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+         Scalar maxval,
+         std::vector<Scalar>& shape,
+         PrimDataType dtype) -> Tensor {
+        FUSER_PERF_SCOPE("Operators.uniform");
```
Comment (kevinstephano) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:2655]: Could you add this check here:

```
        TORCH_CHECK(
            self.validUse(), "Attempting to add to a completed definition!");
```

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 5]
```diff
@@ -2586,6 +2586,72 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+      py::arg("dtype") = DataType::Float,
+      py::return_value_policy::reference);
+  nvf_ops.def(
+      "normal",
+      [](FusionDefinition::Operators& self,
+         Scalar mean,
+         Scalar std,
+         std::vector<Scalar>& shape,
+         PrimDataType dtype) -> Tensor {
+        FUSER_PERF_SCOPE("Operators.normal");
```
Comment (kevinstephano) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:2690]: Same here...

Could you add this check here:

```
        TORCH_CHECK(
            self.validUse(), "Attempting to add to a completed definition!");
```

**Review by kevinstephano — APPROVED** [review]
Could you just add the two checks?  Otherwise, it looks good to go!

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp (hunk 2 above)
Comment (jacobhinkle) [reply third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:2648]: Unless we want `randn_like`...

Comment (jjsjann123) [reply third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:2648]: Yeah, we can revisit that at a later point~ for now I'm just going to enjoy the convenience of the new API :)

**Comment by jjsjann123** [comment]
Tests passed, except the failing CI build against upstream/master. I"m merging this PR.

**Pull request merged**